React Quickstart using new design and auth0-spa-js (#113)
auth0-samples/auth0-react-samples: 01-Login/README.md

# Scenario #1 - Logging In and Gated Content

This sample demonstrates:

- Logging in to Auth0 using Redirect Mode
- Accessing profile information that has been provided in the ID token
- Gated content. The `/profile` route is not accessible without having first logged in

## Project setup

```bash
npm install
```

### Configuration

The project needs to be configured with your Auth0 domain and client ID in order for the authentication flow to work.

> To do this, first copy `src/auth_config.sample.json` into a new file in the same folder called `scr/auth_config.json`, and replace the values within with your own Auth0 application credentials:

Reviewer: > `src/auth_config.sample.json`

The file is actually `src/auth_config.json.example`

Reviewer: > `scr/auth_config.json`

Typo here (should be `src/auth_config.json`)

## License

This project is licensed under the MIT license. See the [LICENSE](../LICENSE) file for more info.

# Tutorial
This tutorial will help you get up and running the React app using the new SDK.

## Setting Up the Application

> 

Reviewer: Let's put the steps in here for running `create-react-app` so that it's clear for the reader how to  get started from scratch.

i.e. "Find a location on your drive, run `<npm command here>`, etc.

### Installing initial dependencies

> After create a new React app using `create-react-app` we still need to install some dependencies that doesn't come with the boilerplate. These dependencies will help us with things like routing, styling and font icons. But, before that we need to move the current `dependencies` to `devDependencies` this is because we will be using a server to run the built app.

Reviewer: I would actually remove the need for the reader to move everything into `devDependencies`. I know we do that for the sample but I'm not sure it's a detail that the reader has to worry about unless they're going to be running this app in production.

Also let's drop the allusions to the styling and fonts, as that's a detail specific to the sample app only (see a later comment about the dependencies).

```suggestion
After create a new React app using `create-react-app` we still need to install `react-router`, which doesn't come as standard with the boilerplate project.
```

> ```json

Reviewer: As per the previous comment, let's take this code sample out.

- "dependencies": {
+ "devDependencies": {
    "react": "^16.8.6",
    "react-dom": "^16.8.6",
    "react-scripts": "2.1.8"
  }
```

> After that we can install the new dev dependencies.

Reviewer: You might want to remove this sentence and just let it run on from the previous one before the code sample.

```bash

> npm i -D react-router-dom highlight.js reactstrap samples-bootstrap-theme \

Reviewer: For the tutorial, the visual aspects of the application can be dropped as it's a detail that's specific to the sample application.

Here, `samples-bootstrap-theme`, `highlight.js`, `reactstrap` and the FontAwesome dependencies can be dropped.

@fortawesome/fontawesome-svg-core @fortawesome/free-solid-svg-icons @fortawesome/react-fontawesome
```
### Create the Home view and initial components

> We are going to create a serie of components in order to build our initial welcome page. Create a folder called `views` and then a Compoment called `Home.js`. This component will show a welcome message and some information. 

Reviewer: ```suggestion
We are going to create a series of components in order to build our initial welcome page. Create a folder called `views` and a compoment called `Home.js` inside that. This component will show a welcome message and some information:
```

```jsx
// src/views/Home.js
import React, { Fragment } from "react";

import Hero from "../components/Hero";
import Content from "../components/Content";

const Home = () => (
  <Fragment>
    <Hero />
    <hr />
    <Content />
  </Fragment>
);

export default Home;
```

> As you can see our `Home.js` component is build with other component pieces. Create a folder called `Compoments` and inside create two new components called `Hero.js`, and `Content.js`.

Reviewer: ```suggestion
As you can see, our `Home.js` component is built with other component pieces.

Next, create a folder called `components` alongside the `views` folder, and inside create another component called `Hero.js`:
```

(See later comment about removing `Content.js`

> For our `Hero.js` compoment we need to move the `logo.svg` inside a folder called `assets`.

Reviewer: ```suggestion
For our `Hero.js` compoment we need to move the `logo.svg` inside a folder called `assets`:
```

Reviewer: Can we have a little more detail here? Where does `logo.svg` come from? Is it already included with the project or should the reader download it from somewhere?

```jsx
// src/components/Hero.js
import React from "react";

import logo from "../assets/logo.svg";

const Hero = () => (
  <div className="text-center hero">
    <img className="mb-3 app-logo" src={logo} alt="React logo" />
    <h1 className="mb-4">React.js Sample Project</h1>

    <p className="lead">
      This is a sample application that demonstrates an authentication flow for
      an SPA, using <a href="https://reactjs.org">React.js</a>
    </p>
  </div>
);

export default Hero;
```

> Since we want to load the content more dinamically, we are going to create a file called `contentData.js` inside a new folder called `utils`. This data will be used inside the `Content.js` component.

Reviewer: This is a detail that is specific to the sample application and not something the reader of the quickstart should really deal with. For the tutorial, it's fine to focus on the authentication parts and not worry about the visual styling and content at all.

So basically this paragraph, the following code snippet and the content component could be dropped from the tutorial. If the reader wants to check out how that is done, they can refer to the sample.

```js
// utils/contentData.js

> const contentData = [

Reviewer: See previous comment about removing this component.

  },
  {
    title: "Enable Multifactor Authentication",
    link: "https://auth0.com/docs/multifactor-authentication",
    description:
      "Add an extra layer of security by enabling Multi-factor Authentication, requiring your users to provide more than one piece of identifying information. Push notifications, authenticator apps, SMS, and DUO Security are supported."
  },
  {
    title: "Anomaly Detection",
    link: "https://auth0.com/docs/anomaly-detection",
    description:
      "Auth0 can detect anomalies and stop malicious attempts to access your application. Anomaly detection can alert you and your users of suspicious activity, as well as block further login attempts."
  },
  {
    title: "Learn About Rules",
    link: "https://auth0.com/docs/rules",
    description:
      "Rules are JavaScript functions that execute when a user authenticates to your application. They run once the authentication process is complete, and you can use them to customize and extend Auth0's capabilities."
  }
];

export default contentData;
```

> ```jsx

Reviewer: See previous comment about removing this component.

            {contentData.map((col, i) => (
              <Col key={i} md={5}>
                <h6 className="mb-3">
                  <a href={col.link}>
                    <FontAwesomeIcon icon="link" />
                    {col.title}
                  </a>
                </h6>
                <p>{col.description}</p>
              </Col>
            ))}
          </Row>
      </div>
    );
  }
}

export default Content;

```

Another component that we need to create for our welcome page is the is the `Footer.js` component:

```jsx
// src/components/Footer.js
import React from "react";

const Footer = () => (
  <footer>
    <div className="logo" />
    <p>
      Sample project provided by <a href="https://auth0.com">Auth0</a>
    </p>
  </footer>
);

export default Footer;
```

### Editing the App component
At the moment we have created our initial components but we need to use them. For that, we need to modify our `App.js` to look like the following code.

```jsx
// src/App.js
import React, { Component, Fragment } from "react";
import { Container } from "reactstrap";

import Footer from "./components/Footer";
import Home from "./views/Home";

> // styles

Reviewer: Styles and fonts can be removed.

import "samples-bootstrap-theme/dist/css/auth0-theme.css";
import "./App.css";

> // fontawesome

Reviewer: Styles and fonts can be removed.

export default function withAuthentication(WrappedComponent, auth0) {
  return class WithAuthentication extends Component {
    state = { loading: true, isAuthenticated: false };

>     async componentWillMount() {

Author: @elkdanger What reason behind the change? according to React's documentation, this will be deprecated and it shouldn't be used anymore: https://reactjs.org/docs/react-component.html#unsafe_componentwillmount

Reviewer: @glrodasz Oops, that's a mistake. I think I was trying to fix something out and forgot to change it back. I'll fix that :)
